pkg/wekafs/dirvolume.go

```diff
@@ -0,0 +1,388 @@
+	}
+	// Available blocks * size per block = available space in bytes
+	maxCapacity := int64(stat.Bavail * uint64(stat.Bsize))
+	glog.V(4).Infoln("Max capacity available for a volume:", maxCapacity)
+	return maxCapacity, nil
+}
+
+func (v DirVolume) GetType() VolumeType {
+	return VolumeTypeDirV1
+}
+
+func (v DirVolume) GetCapacity(mountPath string) (int64, error) {
+	glog.V(3).Infoln("Attempting to get current capacity of volume", v.GetId())
+	if v.apiClient != nil && v.apiClient.SupportsQuotaDirectoryAsVolume() {
+		size, err := v.getSizeFromQuota(mountPath)
+		if err == nil {
+			glog.V(3).Infoln("Current capacity of volume", v.GetId(), "is", size, "obtained via API")
+			return int64(size), nil
+		}
+	} else {
+		size, err := v.getSizeFromXattr(mountPath)
+		if err != nil {
+			return 0, err
+		}
+		glog.V(3).Infoln("Current capacity of volume", v.GetId(), "is", size, "obtained via Xattr")
+		return int64(size), nil
+	}
+	size, err := v.getSizeFromDf(mountPath)
+	if err == nil {
+		glog.V(3).Infoln("Current capacity of volume", v.GetId(), "is", size, "obtained via fs stat")
+		return int64(size), nil
+	}
+	return 0, errors.New("could not fetch volume capacity")
+}
+
+func (v DirVolume) UpdateCapacity(mountPath string, enforceCapacity *bool, capacityLimit int64) error {
+	glog.V(3).Infoln("Updating capacity of the volume", v.GetId(), "to", capacityLimit)
+	f := func() error { return v.updateCapacityQuota(mountPath, enforceCapacity, capacityLimit) }
+	if v.apiClient == nil {
+		glog.V(4).Infof("Volume has no API client bound, updating capacity in legacy mode")
+		f = func() error { return v.updateCapacityXattr(mountPath, enforceCapacity, capacityLimit) }
+	} else if !v.apiClient.SupportsQuotaDirectoryAsVolume() {
+		glog.V(4).Infoln("Updating quota via API not supported by Weka cluster, updating capacity in legacy mode")
+		f = func() error { return v.updateCapacityXattr(mountPath, enforceCapacity, capacityLimit) }
+	}
+	err := f()
+	if err == nil {
+		glog.V(3).Infoln("Successfully updated capacity for volume", v.GetId())
+	}
+	return err
+}
+
+func (v DirVolume) updateCapacityQuota(mountPath string, enforceCapacity *bool, capacityLimit int64) error {
+	if enforceCapacity != nil {
+		glog.V(4).Infoln("Updating quota on volume", v.GetId(), "to", capacityLimit, "enforceCapacity:", *enforceCapacity)
+	} else {
+		glog.V(4).Infoln("Updating quota on volume", v.GetId(), "to", capacityLimit, "enforceCapacity:", "RETAIN")
+	}
+	inodeId, err := v.getInodeId(mountPath)
+	if err != nil {
+		glog.Errorln("Failed to fetch inode ID for volume", v.GetId())
+		return err
+	}
+	fs, err := v.getFilesystemObj()
+	if err != nil {
+		glog.Errorln("Failed to fetch filesystem for volume", v.GetId())
+		return err
+	}
+
+	q, err := v.apiClient.GetQuotaByFileSystemAndInode(fs, inodeId)
+	// check if the quota already exists. If not - create it and exit
+	if err != nil {
+		if err != apiclient.ObjectNotFoundError {
```

Comment:
you could use this for falling back to xattr if quota not found, i.e legacy volue

Reply:
yes. did it on updateCapacity flow but not in GetCapacity